### ISSUE TITLE
viewport switching with "Desktop wall" plugin is laggy

### ISSUE THREAD

**Issue by rabin-io** [compiz-reloaded/compiz#120]
when I move between desktops (with and w/o a window) the animations fills a bit "laggy", unlike other animations I enabled (wobbly windows and other window animation from the Animations plugin)

- I'm running Fedora 27 
- on a Lenovo p50 with nvidia Quadro M2000M, 
- using the drivers from negativo17 repo.

(what else i need to provide to debug this issue ?)

**Comment by soreau** [comment]
@rabin-io Hi, it might be that the animation in question is damaging both screens (an expensive operation) when it should only be damaging the area of animation. To see, you can use the [showrepaint plugin](https://github.com/soreau/showrepaint) which shows colored areas where the screen is repainted. Build it with `make && make install`, remove it with `make uninstall` If both screens are flashing while the animation is happening with showrepaint enabled, this is the case.

**Comment by rabin-io** [comment]
Hi @soreau thanks for the reply, but I have few questions,

I'm using compiz directly from fedora repos (i'm using the MATE spin), 
do I need to compile all compiz or just the plugin ?

what do you mean by "damaging both screens" are you referring to the 2 desktops ?
or to actual physical (external) displays ?

**Comment by soreau** [comment]
You only have to build the plugin. It installs to ~/.compiz/ where compiz and ccsm look first for plugins to load.

When you said move between desktops, I thought you meant two physical screens but I see now I misunderstood. The wall plugin animates the desktop switches, the algorithm can be found in the [wall plugin code](https://github.com/compiz-reloaded/compiz/blob/master/plugins/wall.c). The showrepaint plugin would show that all of the screen is being damaged during desktop switches but only relevant parts of the screen are damaged on animations like wobbly. You might notice that small windows are smoother than large windows when it comes to wobbly. What is your resolution? Higher resolutions can take longer to process. It could also be a feature specific to the version of the driver you're using.

**Comment by rabin-io** [comment]
OK, compiled and installed the plugin

    dnf install -y compiz-devel compiz-bcop.noarch libtool mesa-libGL-devel

I enabled it (its a bit a mess) and it seems like all my desktop is constantly is updating, 
one thing I notice is that while doing the desktop switch using the keybinding (in my case i use Ctrl+Alt+⭠/⭢) is while still holding the keys (Ctrl+Alt) while nothing is chaning and i'm not moving, 
the screen is re-painted in a very high frequency (even when disabling=0 the wall sliding duration ).

**Comment by soreau** [comment]
What do your fps look like with benchmark plugin? Maybe you should try toggling sync to vblank in general options.

**Comment by soreau** [comment]
Well the fact that you say it seems the desktop is constantly updating.. can you try disabling each plugin one at a time to see which is causing the constant refreshing?

**Comment by rabin-io** [comment]
Not all of  my desktop is constantly updating (just some parts of it) because I use the system resource monitor which plot the system resource on the top panel. and I also have a network traffic applet which constantly is updating.

The part where I switch virtual desktops and still holding the trigger keys and the screen is re-painted is a bug, not sure how to define it, but is seems while I still hold the "keys" i'm still allowed to move between desktops and while holding them there seems to trigger full screen repaint.

**Comment by soreau** [comment]
This is the point where you'd likely start debugging, by reading the code or placing some printf statements or using gdb from another machine ssh'd in. The goal being to find out why holding those modifiers causes fullscreen repaints. Does the repainting start instantly when you press the modifiers or only after switching desktops? This could be a clue to help isolate the problem.

**Comment by rabin-io** [comment]
I didn't touch `C` for ages, and the lack of understanding how a window manager works is also an issue to debug this, but my guess would be that code relvent to the problem is around [line 1490](https://github.com/compiz-reloaded/compiz/blob/master/plugins/wall.c#L1490) where i'm not moving but still holding the keys, 

or something about this lines, maybe
```C
    UNWRAP (ws, s, preparePaintScreen);
    (*s->preparePaintScreen) (s, msSinceLastPaint);
    WRAP (ws, s, preparePaintScreen, wallPreparePaintScreen);
```

**Comment by soreau** [comment]
I can reproduce this but it doesn't really cause a slowdown on my system (using amdgpu+radeonsi). In the code, I would guess the call to damageScreen() is being called in wallDonePaintScreen() due to one of the cases being true there. If this is the case, it would be a matter of figuring out which are true and why, then make it so damageScreen() isn't being called every frame while just holding the modifier keys.

**Comment by rabin-io** [comment]
I don't think this issue is related to the OP issue, this is just another "not so ideal behaviour" I found out with the plugin. 

for now I removed the animation so that skip from one desktop to another be fast.

@soreau, do you have any other pointers how to investigate the transition lag ?

**Comment by soreau** [comment]
I put some debug code in and it looks like ws->showPreview and ws->boxTimeout are both true, causing damageScreen() to be called where I expected. showPreview means what it says, the preview rendering the desktop layout in the middle of the screen. boxTimeout is used for fading out the preview, after releasing the bindings.

Disabling 'Show Viewport Switcher Preview' indeed eliminates the screen refreshing. This could be optimized to only damage the actual region of the switcher preview, and stop damaging when the pixels stop changing. Does disabling this option help the performance in your case?

**Comment by rabin-io** [comment]
No, I never had this open enabled.

**Comment by soreau** [comment]
Without that option enabled, I am unable to reproduce the refreshing problem, so I'm kinda out of ideas on this one for now.
